### ISSUE TITLE
haproxy 2.0.5 : http client_timeout trigger with "CD--" instead "cD--" in logs with http2 frontend only

### ISSUE THREAD

**Issue by aderumier** [haproxy/haproxy#308]
Hi,

I was debugging a customer problem, trying to POST a form.

The application is sending fast the content of the POST, but waiting between 30-60s before getting first byte of result from the server.

timeout configuration:

        timeout connect 5s
        timeout check 3s
        timeout queue 30s
        timeout client 30s
        timeout server 300s
        timeout http-request 10s
        timeout http-keep-alive 10s

After 30s, the client is disconnect with this logs
FO~ xxx/xxxxx 0/0/1/-1/30003 -1 0 - - CD-- 1/1/0/0/0 0/0  "POST /xxxx HTTP/1.1"

reducing timeout client to 15s

 FO~ xxx/xxxx 0/0/1/-1/15004 -1 0 - - CD-- 1/1/0/0/0 0/0 "POST /xxxx HTTP/1.1"

increasing the timeout client to 60s ->ok

FO~ xxx/xxxx 0/0/1/33178/33179 200 46345 - - ---- 1/1/0/0/0 0/0 "POST /xxxx HTTP/1.1"

Is it normal to have "CD" instead "cD" ?

This is with haproxy 2.0.5, with ot without htx.

**Comment by aderumier** [comment]
After testing, it seem to occur only with http2 (client->frontend http2).

when disabling http2 on the frontend,
setting "client timeout 15s", the query works fine after 34s

 FO~ xxx/xxxxx 0/0/1/34831/34889 200 46354 - - ---- 1/1/0/0/0 0/0 "POST /xxxx HTTP/1.1"

**Comment by aderumier** [comment]
curl request http1.1
```
curl 'https://xxxxxxxxxx' -H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.14; rv:69.0) Gecko/20100101 Firefox/69.0' -H 'Accept: application/ld+json' -H 'Accept-Language: fr,fr-FR;q=0.8,en-US;q=0.5,en;q=0.3' --compressed -H 'Content-Type: application/ld+json' -H 'Authorization: xxxx' -H 'Origin: https://delivery-bo-mtgt.m6tech.net' -H 'Connection: keep-alive' -H 'Referer: https://xxxxx' --data $'......'

```
curl request http2
```
curl 'https://xxxxx/xxxxx' -H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.14; rv:69.0) Gecko/20100101 Firefox/69.0' -H 'Accept: application/ld+json' -H 'Accept-Language: fr,fr-FR;q=0.8,en-US;q=0.5,en;q=0.3' --compressed -H 'Content-Type: application/ld+json' -H 'Authorization: xxxxx' -H 'Origin: https://xxxxx' -H 'Connection: keep-alive' -H 'Referer: https://xxxxx' -H 'TE: Trailers' --data $'......'
```

The only difference is -H 'TE: Trailers'  with http2

**Comment by wtarreau** [comment]
Hi Alexandre,

I suppose there are multiple issues in your report, but it's a bit confusing
to me :
  - timeout reported after the request was fully sent
  - CD instead of cD reported on timeout

Am I right ? If so I think it could be related to the client timeout striking
on the H2 connection, causing it to be closed, and reporting the error. I'm
realizing that from the H2 perspective it's not easy to know if some streams
are actively waiting for the other side or not, so I'll go back check the
code to see if that's properly handled. I think we'd need to check if there
are pending data in the output buffer before deciding to consider timeouts.

However it's true that we'd need to enforce a longer timeout on the connection
than the one on the stream otherwise we can kill the connection at the lower
layer first.

**Comment by aderumier** [comment]
Hi Willy,

>>I suppose there are multiple issues in your report, but it's a bit confusing
>>to me :
>>  - timeout reported after the request was fully sent
>>  - CD instead of cD reported on timeout



yes, sorry. 
My main concern was that "timeout client" don't seem to work like http1. 
Seem that with http2, I need to increase "timeout client" to wait for response from the server.
I really don't known if it's a bug or if it's expected with http2.


I'll try to do a debug client side with curl to have more info.
I can reproduce it 100%.

**Comment by wtarreau** [comment]
Could you please try the attached patch ? It should address this. It's not 100% complete yet as we don't have a list of streams waiting for a WINDOW_UPDATE so we could theorically face some non-expiring streams in rare situations but overall it should be ok.

Arrrggh, shitty github file uploader! Let's try again!
[0001-WIP-BUG-MEDIUM-mux-h2-do-not-enforce-timeout-on-long.patch.txt](https://github.com/haproxy/haproxy/files/3675169/0001-WIP-BUG-MEDIUM-mux-h2-do-not-enforce-timeout-on-long.patch.txt)

**Comment by aderumier** [comment]
It's working ! (patch applied on 2.0.7).

Thanks Willy !

**Comment by wtarreau** [comment]
Great, thanks Alexandre. Now I need to figure how to better address
the remaining case of flow-controlled streams.

**Comment by wtarreau** [comment]
Alexandre, I could finally address this one entirely and verify that it works fine. Now it properly takes care of flow-controlled streams as well. It's not backported yet, there are 3 patches to take in total (all mentioned in the commit if you're interested).

**Comment by aderumier** [comment]
I'm going to test it next week :)

Thanks again Willy for the fast patch! (I'll try to go to amsterdam conf, I'll offer you a beer ;)

**Comment by wtarreau** [comment]
The 1.8 backport requires too many changes to the fragile H2 code. I'll leave it as wontfix for 1.8.

**Comment by TreZc0** [comment]
Sorry for stumbling into this ticket, but i just stumbled across "CD--" in my logs after migrating to http/2 for frontend and backend on Haproxy v2.2.9 and have a hard time finding out if this was encorporated into the main branch for HaProxy 2.2 release, as the commits displayed in the last message are not in the main repo even today.

Could someone please clarify this for me?

**Comment by capflam** [comment]
The patch (c2ea47f) was merged during the 2.1 development. Thus it is also part of the 2.2.
